### PR TITLE
Better deploy results handling for stream responses

### PR DESCRIPTION
**Pull request by vshlapakov**
In some cases deploy can stream responses, and in most such cases you don't want to get full deploy logs each time. A better way could be checking last line and if it has an error status, then show last N rows from the log (N=30).

Review, please.

[Review on Reviewable](https://reviewable.io/reviews/scrapinghub/shub/96)

### CODE REVIEW

**Comment by dangra** [comment]
Instead, it can store the response in a temporal file and show the last N lines on failures, but also link to the complete logfile in case users want to see it.

**Comment by vshlapakov** [comment]
@dangra Good idea, something like that?

File: shub/utils.py [hunk 1]
```diff
@@ -47,6 +61,36 @@ def make_deploy_request(url, data, files, auth):
...
+        if 'status' in data and data['status'] == 'ok':
+            return True
+    return False
+
+
+def _log_bad_deploy():
+    if not os.path.exists(DEPLOY_LOG):
+        return
+    logs = subprocess.check_output(
+        ['tail', '-n', str(LAST_N_LOGS), DEPLOY_LOG])
```
Comment (pablohoffman) [inline shub/utils.py]: it needs to work on windows. `shub` is cross platform

Comment (vshlapakov) [inline shub/utils.py]: Sure, sorry, simplified this part and made it cross-platform.

**Comment by pablohoffman** [comment]
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[shub/utils.py, line 42 \[r3\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6dpzzJgczAPEXzi3CG:-K6dpzzJgczAPEXzi3CH:1372143728) ([raw file](https://github.com/scrapinghub/shub/blob/5be85d439e25aaddb095ce9597845f5a33d92ac7/shub/utils.py#L42)):*
why not use `NamedTemporaryFile`?

---

*[shub/utils.py, line 67 \[r3\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6dqh-p5O4O8VhKoPDC:-K6dqh-qi_ES6qlsiHvG:1137741422) ([raw file](https://github.com/scrapinghub/shub/blob/5be85d439e25aaddb095ce9597845f5a33d92ac7/shub/utils.py#L67)):*
a test for this function would be nice. Otherwise, a simpler way to check for the state of a deploy (like checking the HTTP status)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by vshlapakov** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[shub/utils.py, line 67 \[r3\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6dqh-p5O4O8VhKoPDC:-K6g_DC8aGlWQY66loIr:-675136760) ([raw file](https://github.com/scrapinghub/shub/blob/5be85d439e25aaddb095ce9597845f5a33d92ac7/shub/utils.py#L67)):*
Added some basic tests, take a look please. I'm afraid that we can't rely on HTTP status with streaming responses and have to check a last line for status field.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by vshlapakov** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[shub/utils.py, line 42 \[r3\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6dpzzJgczAPEXzi3CG:-K6gnuHkEOLmyNnUBEg9:-190368324) ([raw file](https://github.com/scrapinghub/shub/blob/5be85d439e25aaddb095ce9597845f5a33d92ac7/shub/utils.py#L42)):*
Revised and fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by pablohoffman** [comment]
Reviewed 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[shub/utils.py, line 46 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hacjGSRwwSRmNuL_b:-K6hacjHOs6VgeZX0UAL:158154653) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L46)):*
why is the `flush()`needed on every line?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by vshlapakov** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[shub/utils.py, line 46 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hacjGSRwwSRmNuL_b:-K6hcSlcgyysdc8vUMCD:-185023898) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L46)):*
Well, for real-time logging (only). But yeah, looks like it's not very useful, as we are going to show the log path only in failure cases according to the last changes..Let me remove it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by dangra** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[shub/utils.py, line 42 \[r3\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6dpzzJgczAPEXzi3CG:-K6dtR_Lt4Br9gIiqsNf:1640142104) ([raw file](https://github.com/scrapinghub/shub/blob/5be85d439e25aaddb095ce9597845f5a33d92ac7/shub/utils.py#L42)):*
+1 to NamedTemporaryFile with prefix="shub_deploy" and suffix=".log", also delete=False.

---

*[shub/utils.py, line 46 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hacjGSRwwSRmNuL_b:-K6hehQ1lgc2W05tmhMR:9700251) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L46)):*
I think we want to do both, storing to a file and showing its location in case of error. But also showing the build process while it is happening when using a command line flag.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by dangra** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[shub/utils.py, line 70 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hfQAgUq9IDFNVD4pI:-K6hfQAgUq9IDFNVD4pJ:-266477177) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L70)):*
Last line is json afaik, why is it using `ast.literal_eval()` instead of `json.loads()`? 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by vshlapakov** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[shub/utils.py, line 42 \[r3\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6dpzzJgczAPEXzi3CG:-K6iBzlfdrsSd9LobklK:-499146009) ([raw file](https://github.com/scrapinghub/shub/blob/5be85d439e25aaddb095ce9597845f5a33d92ac7/shub/utils.py#L42)):*
Done.

---

*[shub/utils.py, line 46 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hacjGSRwwSRmNuL_b:-K6iC15PxMkSndBOWgsD:1457700874) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L46)):*
Thanks, done.

---

*[shub/utils.py, line 70 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hfQAgUq9IDFNVD4pI:-K6iBcit-rlUVaXU7auE:-1535081722) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L70)):*
Of course you're right, fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by pablohoffman** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[shub/deploy.py, line 40 \[r6\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6iXzKwFOxCX8y7AD4_:-K6iXzKwFOxCX8y7AD4a:-79594729) ([raw file](https://github.com/scrapinghub/shub/blob/c2b0857e1eb8e053f44a0db2f75705e996dc9882/shub/deploy.py#L40)):*
I'd use `-v` for verbose and swap it with `--version`, or remove the shortcut for version altogether.

---

*[shub/utils.py, line 46 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hacjGSRwwSRmNuL_b:-K6iYZkNXJEdDsU5W3Fo:211197472) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L46)):*
Doesn't "doing both" mean showing real-time and store to log?. In which case, perhaps we still need the flushing?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by dangra** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[shub/deploy.py, line 40 \[r6\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6iXzKwFOxCX8y7AD4_:-K6iuQ_7V3snSdtzIL1P:1980499780) ([raw file](https://github.com/scrapinghub/shub/blob/c2b0857e1eb8e053f44a0db2f75705e996dc9882/shub/deploy.py#L40)):*
I agree, +1 to `-v` for `--verbose` and `-V` for `--version`

---

*[shub/utils.py, line 46 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hacjGSRwwSRmNuL_b:-K6iui8jghzwjvh5HBA0:1738663085) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L46)):*
The idea is to:

* Always write to temporal logfile (no need to flush)
* Remove logfile on success build unless `--keep-deploy-log` is passed
* Do not show build process by default on console, just store it into logfile
* Do not remove logfile and show its location in case of build failure
*  Show build process on console when `--verbose` is passed (still store logfile and previous also rules apply)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by pablohoffman** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[shub/utils.py, line 46 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hacjGSRwwSRmNuL_b:-K6iwRt8ByIbsqwuGImO:846065000) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L46)):*
Sounds good. I'd change the option name to `--keep-log`, `keep-deploy-log` is a bit too long and sounds redundante since this is already within the context of a deploy command.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by dangra** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[shub/utils.py, line 46 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hacjGSRwwSRmNuL_b:-K6iy3LGLWSIClmPCCX5:1425820692) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L46)):*
`--keep-log` is fine by me.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by pablohoffman** [comment]
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by vshlapakov** [comment]
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[shub/deploy.py, line 40 \[r6\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6iXzKwFOxCX8y7AD4_:-K6jHq5s8BenTRFTq0My:1935484215) ([raw file](https://github.com/scrapinghub/shub/blob/c2b0857e1eb8e053f44a0db2f75705e996dc9882/shub/deploy.py#L40)):*
Swapped.

---

*[shub/utils.py, line 46 \[r5\]](https://reviewable.io:443/reviews/scrapinghub/shub/96#-K6hacjGSRwwSRmNuL_b:-K6jHu5h7jB61B9bRDZ1:292614926) ([raw file](https://github.com/scrapinghub/shub/blob/53ecdd82c99c0f068bfc3d4c3dec2ec33f5e4594/shub/utils.py#L46)):*
:+1: Updated the name.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by pablohoffman** [comment]
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96)*

**Comment by pablohoffman** [comment]
Looks OK, can you rebase @vshlapakov ?

**Comment by vshlapakov** [comment]
@pablohoffman Sure, rebased and squashed.

**Comment by pablohoffman** [comment]
thanks. @jdemaeyer can you merge after a final review?. Code looks OK to me. Thanks!

**Comment by jdemaeyer** [comment]
Thanks @vshlapakov

---

Reviewed 1 of 1 files at r9.
Review status: 1 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/scrapinghub/shub/96#-:-K6mW9qsSzpAq7DBqFm1:2026354511)*

**Comment by vshlapakov** [comment]
Thank you all for the help with it!

**Pull request merged**